universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java

```diff
@@ -189,23 +189,38 @@ public ProgramDefinition moveBlock(long blockId, Direction direction)
         blockSlice.startsBefore(otherBlockSlice) ? blockSlice : otherBlockSlice;
     BlockSlice latterSlice =
         blockSlice.startsBefore(otherBlockSlice) ? otherBlockSlice : blockSlice;
+
+    // An illegal move is when a predicate in a later block depends on an earlier block.
+    // Get the questions used in any predicate in the latter block.
+    ImmutableSet<Long> predicateQuestions =
+        blockDefinitions().subList(latterSlice.startIndex(), latterSlice.endIndex()).stream()
+            .filter(b -> b.visibilityPredicate().isPresent())
```

Comment:
(super nit) you could use `Optional.map` instead of `Stream.filter`. Not actually recommending it because it might be _less_ readable...

Reply:
I always forget about `Optional.map`, thanks for the reminder! This is no longer relevant with the new method